# Neural Network

We are going to use housing prices dataset to predict the housing prices based on multiple features. <br>
Unlike linear regression where we deal with only two variable x and y, using neural network we can train our model on multiple feature variables.

Getting Required Libraries.

In [1]:
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import seaborn as sns
import keras

Using TensorFlow backend.


# Getting our dataset into a pandas dataframe using read_csv() method.

In [2]:
data = pd.read_csv('HousingPrices.csv')

Using head() we can see the first 5 rows of our dataset, this gives us ann about about our data without going into all the lines.

In [3]:
data.head()

YearBuilt  2ndFlrSF  GrLivArea  FullBath  HalfBath  BedroomAbvGr  \
0       2003       854       1710         2         1             3   
1       1976         0       1262         2         0             3   
2       2001       866       1786         2         1             3   
3       1915       756       1717         1         0             3   
4       2000      1053       2198         2         1             4   

   TotRmsAbvGrd  YrSold  SalePrice  
0             8    2008     208500  
1             6    2007     181500  
2             6    2008     223500  
3             7    2006     140000  
4             9    2008     250000

# Getting our variables for training.<br>
x will contain all the values in a row except for SalePrice, and our model is going to be trained on these values <br>
y will only contain SalePrice, which we are gonna predict

In [4]:
x = data.drop(columns=['SalePrice'])

In [5]:
y = data.SalePrice

# Model Building


In [6]:
model = keras.Sequential() # Created a sequential model object
model.add(keras.layers.Dense(8, activation='relu', input_shape=(8,))) #our first layer is an input layer
model.add(keras.layers.Dense(8,activation='relu')) #Hidden layer
model.add(keras.layers.Dense(1)) # output layer

Description of layers:<br>
1st layer contains 8 neurons, i have taken 8 neurons because in dataset i have 8 features to train on, but you can choose any number if you want. I have used RELU as an activation function because it is mostly used, input shape is shape of our dataset, we have 8 columns here.<br>
2nd layes is hiddel layer, it is an addition layer. You can have as many hidden layers as you want.<br>
3rd layer is our output layer, here we are gonna predict only SalePrice thats why 1 is given in argument.


# Now we will need to Compile our model.

i have used adam optimizer but there are other optimizers too.<br>
loss is nothing but the error our model while predicting some output.

In [7]:
model.compile(optimizer='adam', loss='mean_squared_error')

## Finally Fit() our model.<br>
Callbacks are used to avoid overfitting. <br>
using early stopping, we can stop the model training when it is showing same loss between the epochs. This saves memory and time to train.

In [8]:
model.fit(x,y,epochs=50, callbacks=[keras.callbacks.EarlyStopping(patience=5)])

Epoch 1/50
1460/1460 [==============================] - 1s 681us/step - loss: 39043831179.5726
Epoch 2/50
1460/1460 [==============================] - 0s 126us/step - loss: 38599885944.6356
Epoch 3/50


/home/sanket/anaconda3/lib/python3.7/site-packages/keras/callbacks/callbacks.py:846: RuntimeWarning: Early stopping conditioned on metric `val_loss` which is not available. Available metrics are: loss
  (self.monitor, ','.join(list(logs.keys()))), RuntimeWarning


1460/1460 [==============================] - 0s 152us/step - loss: 37867991118.5534
Epoch 4/50
1460/1460 [==============================] - 0s 310us/step - loss: 36780659776.5260
Epoch 5/50
1460/1460 [==============================] - 0s 170us/step - loss: 35217906180.2082
Epoch 6/50
1460/1460 [==============================] - 0s 123us/step - loss: 33087535541.6548
Epoch 7/50
1460/1460 [==============================] - 0s 130us/step - loss: 30362277169.7973
Epoch 8/50
1460/1460 [==============================] - 0s 119us/step - loss: 27093241367.8466 0s - loss: 27652795549.53
Epoch 9/50
1460/1460 [==============================] - 0s 76us/step - loss: 23409548481.5781
Epoch 10/50
1460/1460 [==============================] - 0s 175us/step - loss: 19531553340.3178
Epoch 11/50
1460/1460 [==============================] - 0s 176us/step - loss: 15721433550.9041
Epoch 12/50
1460/1460 [==============================] - 0s 175us/step - loss: 12237211642.3890
Epoch 13/50
1460/1460 [==========

# Testing our Model

So lets pass some values to the predict method of our model.<br>
Basically, here i have copied all the values from 1st collumn of the dataset, and used predict method using those values.

In [9]:
data[40:41]

YearBuilt  2ndFlrSF  GrLivArea  FullBath  HalfBath  BedroomAbvGr  \
40       1965         0       1324         2         0             3   

    TotRmsAbvGrd  YrSold  SalePrice  
40             6    2006     160000

In [10]:
test_data = np.array([1965,0,1324,2,0,3,6,2006]) # Passing some values to predict
print(model.predict(test_data.reshape(1,8), batch_size=1))  #we need to reshape to arrange all the data in one row

[[166042.19]]


Our model has predicted that SalePrice is near to 166042, when actually it is 160000. So its not accurate but pretty close actually

# Saving the model for later use

In [11]:
model.save('housing_model.h5')

# Loading Previously saved model

In [13]:
old_model = keras.models.load_model('housing_model.h5')

In [14]:
old_model.predict(test_data.reshape(1,8))

array([[166042.19]], dtype=float32)